### PR TITLE
Add depth-first and breadth-first traversal

### PR DESCRIPTION
**Pull request by chaosmail**

### CODE REVIEW

**Comment by chaosmail** [comment]
@pyu10055 i pushed the code for both DFS and BFS. Maybe you would like to take a quick look. I re-added the parameter `i` in addition to the new depth parameter in the `iteratorFn`, because I find both very useful.

**Comment by pyu10055** [comment]
Looks good, thanks

File: src/dag.ts [hunk 1]
```diff
@@ -32,14 +42,54 @@ const arrToObj = (d: any[], attr: string) => d.reduce((r: any, c: any) => {
...
+ * Traverse a Directed Acyclic Graph via Depth-first search
+ * @arg nodes array of nodes describing the graph
+ * @arg edges array of edges describing the graph
+ * @arg func iterator callback
+ * @arg untilNode iterate until specific node
+ */
+export function iterateDfs<T>(nodes: INode<T>[], edges: IEdge[], func: IteratorFn<T>, untilNode?: string) {
+  const options: IIteratorOptions = {
+    traversalMode: TraversalMode.DFS,
+    untilNode: untilNode,
```
Comment (pyu10055) [inline src/dag.ts]: this line can be just untilNode, similar in the bfs method

Comment (chaosmail) [reply src/dag.ts]: cool, I didn't know you can do this. However, I think here using `{key: val}` is more verbose and avoids the error of ordering the properties.

Comment (pyu10055) [reply src/dag.ts]: sounds good.

**Pull request merged**